### ISSUE TITLE
[Docs] How to efficiently sample from posterior?

### ISSUE THREAD

**Issue by LemonPi** [cornellius-gp/gpytorch#1066]
# 📚 Documentation/Examples
** Is there a feature that needs some example code? **
All of the tutorials/notebooks I've seen stop at constructing the posterior and plotting its mean and confidence region. I'd like to know how to efficiently sample from this posterior. My use case is that I have few training points (~100) but must query large batches at a time (~10000). It doesn't seem like `likelihood(model(test_x)).sample()` is the best way to do this because it scales super-linearly with the size of `test_x` and sampling takes about 10x longer than constructing the posterior (might have something to do with lazy tensors). What's the best way to sample from the posterior?

The code below is a slight modification of the Batch Independent Multioutput GP [tutorial](https://gpytorch.readthedocs.io/en/latest/examples/03_Multitask_Exact_GPs/Batch_Independent_Multioutput_GP.html)

With 1000 test points I get
`posterior 0.0480 mean 0.0000 sample 0.2009`
With 10000 test points I get
`posterior 1.2291 mean 0.0000 sample 4.9208`
Which is more than 10x from before. Why am I getting super-linear scaling? I'd like for the test points to be sampled independently of each other (basically just construct the posterior from the training points and sample test points iid), which should give linear scaling at most - is there a good way of doing this?

```python
import math
import torch
import gpytorch
from matplotlib import pyplot as plt

train_x = torch.linspace(0, 1, 100)
ntest = 10000

train_y = torch.stack([
    torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.shape[0]) * 0.2,
    torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.shape[0]) * 0.8,
], -1)
ny = train_y.shape[1]


class BatchIndependentMultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean(batch_shape=torch.Size([ny]))
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([ny])),
            batch_shape=torch.Size([ny])
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal.from_batch_mvn(
            gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
        )


likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=ny)
model = BatchIndependentMultitaskGPModel(train_x, train_y, likelihood)

training_iterations = 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam([
    {'params': model.parameters()},  # Includes GaussianLikelihood parameters
], lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(training_iterations):
    optimizer.zero_grad()
    output = model(train_x)
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
    optimizer.step()
# Set into eval mode
model.eval()
likelihood.eval()

# Make predictions
with torch.no_grad():
    import time
    start = time.time()
    test_x = torch.linspace(0, 1, ntest)
    predictions = likelihood(model(test_x))
    build_posterior_time = time.time()
    mean = predictions.mean
    get_mean_std_time = time.time()
    sample = predictions.sample()
    final_time = time.time()
    print("posterior {:.4f} mean {:.4f} sample {:.4f}".format(build_posterior_time - start, get_mean_std_time - build_posterior_time, final_time - get_mean_std_time))
```

**Comment by LemonPi** [comment]
One solution I found is to construct a batch normal distribution manually:
```python
    px = torch.distributions.Normal(predictions.mean, predictions.stddev)
    sample2 = px.sample()
```

So the final prediction block becomes
```python
with torch.no_grad():
    import time

    start = time.time()
    test_x = torch.linspace(0, 1, ntest)
    predictions = likelihood(model(test_x))
    build_posterior_time = time.time()
    mean = predictions.mean
    get_mean_std_time = time.time()
    sample = predictions.sample()
    final_time = time.time()

    # manual sampling
    px = torch.distributions.Normal(predictions.mean, predictions.stddev)
    sample2 = px.sample()

    actual_final_time = time.time()
    print("posterior {:.4f} mean {:.4f} sample {:.4f} manual sample {:.4f}".format(build_posterior_time - start,
                                                                                   get_mean_std_time - build_posterior_time,
                                                                                   final_time - get_mean_std_time,
                                                                                   actual_final_time - final_time))
```

Which gives `posterior 0.6545 mean 0.0000 sample 4.6117 manual sample 0.0095` for 10000 test points.

Note that I had to replace the GP model with the multitask one since there's issue #1065 with the BatchIndependentMultitaskGPModel

**Comment by gpleiss** [comment]
> Why am I getting super-linear scaling?

This is because you are trying to sample from the joint distribution of all test points. This requires taking the Cholesky decomposition (or, more generally, the Lanczos decomposition) which scales as O(n^3) ( or O(n^2), respectively ).

> One solution I found is to construct a batch normal distribution manually

Given that you want to sample each point independently, I think this is the best option.

**Comment by LemonPi** [comment]
Thanks!
